asciidoctorj-core/src/main/java/org/asciidoctor/jruby/internal/JavaExtensionRegistryImpl.java

```diff
@@ -318,12 +324,27 @@ public JavaExtensionRegistry inlineMacro(String inlineMacroProcessor) {
         }
     }
 
-    private String getName(Class<?> clazz) {
-        Name nameAnnotation = clazz.getAnnotation(Name.class);
-        if (nameAnnotation == null) {
-            throw new IllegalArgumentException(clazz + " must be registered with a name or it must have a Name annotation!");
+    private String getName(List<Annotation> annotations) {
+        String name = null;
+        Optional<Annotation> nameAnnotation = annotations.stream().filter(a -> a.annotationType().isAssignableFrom(Name.class)).findFirst();
+
+        if (nameAnnotation.isPresent()) {
+            name = ((Name)nameAnnotation.get()).value();
+        } else {
+            for (Annotation anno : annotations) {
+                List<Annotation> tempAnnotationList = new ArrayList<>(Arrays.asList(anno.annotationType().getAnnotations()));
+                tempAnnotationList.removeAll(annotations);
```

Comment:
This is to avoid following direct cycles, right? (Class C has an annotation A1 and A2, where A2 also has an annotation A1).
Are there situations where other longer cycles could happen?
What about doing a breadth-first-search that keeps a context of all yet visited annotations?
I think that would be clearer.

Reply:
Hi @robertpanzer,

Some time has passed, but now I found time again to continue here. ;)
I introduce a list to mark all visited annotations, so i think this approach is a better way, than the first one.